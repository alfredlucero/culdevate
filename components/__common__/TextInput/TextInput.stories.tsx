import React from "react";
import { storiesOf } from "@storybook/react";
import StoryFrame from "../StoryFrame";
import TextInput from "./index";

storiesOf("TextInput", module).add("Default", () => (
  <StoryFrame>
    <TextInput value="someTextInput" name="textInputName" valid={true} touched={true} type="text" onChange={() => {}} />
  </StoryFrame>
));
