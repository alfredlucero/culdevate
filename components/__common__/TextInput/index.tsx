import React from "react";
import styled from "styled-components";
import { culdevateDefaultTheme } from "../../defaultTheme";

const StyledTextInputContainer = styled.div`
  display: flex;
  flex-direction: column;
`;
StyledTextInputContainer.defaultProps = {
  theme: culdevateDefaultTheme,
};

interface StyledTextInputProps {
  isError: boolean;
}
const StyledTextInput = styled.input<StyledTextInputProps>`
  width: 100%;
  color: ${props => props.theme.colors.gray};
  padding: 1.5rem;
  border: 2px solid ${props => (props.isError ? props.theme.colors.red : props.theme.colors.lightGray)};
  border-radius: 5px;
`;
StyledTextInput.defaultProps = {
  theme: culdevateDefaultTheme,
};

export type TextInputType = "text" | "password" | "email";

export interface TextInputProps {
  value: string;
  id: string;
  name: string;
  valid: boolean;
  touched: boolean;
  type: TextInputType;
  onChange: (event: React.FormEvent<HTMLInputElement>) => void;
  label?: string;
  disabled?: boolean;
  infoMessage?: string;
  errorMessage?: string;
  placeholder?: string;
  required?: boolean;
}

const TextInput: React.FC<TextInputProps> = ({
  value,
  id,
  name,
  valid,
  touched,
  type,
  onChange,
  label = "",
  disabled = false,
  infoMessage = "",
  errorMessage = "",
  placeholder = "",
  required = true,
  ...passThroughProps
}) => {
  const infoId = infoMessage && `${id}-text-input-info`;
  const errorId = errorMessage && `${id}-text-input-error`;
  const isError = touched && !valid;

  return (
    <StyledTextInputContainer>
      <label htmlFor={id}>
        {label} {label && required && "(Required)"}
      </label>
      <StyledTextInput
        isError={isError}
        type={type}
        value={value}
        onChange={onChange}
        id={id}
        name={name}
        disabled={disabled}
        placeholder={placeholder}
        aria-describedby={isError ? errorId : infoId}
        {...passThroughProps}
      />
      {!isError && (
        <span data-testid="textInputInfo" id={infoId}>
          {infoMessage}
        </span>
      )}
      {isError && (
        <span data-testid="textInputError" id={errorId}>
          {errorMessage}
        </span>
      )}
    </StyledTextInputContainer>
  );
};

export default TextInput;
