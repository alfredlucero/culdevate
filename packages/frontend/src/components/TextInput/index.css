.text-input {
  @apply shadow appearance-none border rounded w-full py-2 px-3 text-gray-700 leading-tight;
}

.text-input:focus {
  @apply outline-none shadow-outline;
}

.text-input-error {
  @apply border-red-500;
}

.text-input-disabled {
  @apply bg-gray-200 opacity-75;
}

.text-input-label {
  @apply block text-gray-700 text-sm font-bold mb-2;
}

.text-input-info {
  @apply text-gray-700 text-xs italic mt-2;
}

.text-input-error-info {
  @apply text-red-500 text-xs italic mt-2;
}
