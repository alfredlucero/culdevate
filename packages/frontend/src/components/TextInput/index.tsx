import React from "react";
import cn from "classnames";
import { CommonProps } from "../commonProps";
import "./index.css";

interface TextInputProps extends CommonProps {
  onChange: (e: React.FormEvent<HTMLInputElement>) => void;
  value: string;
  id: string;
  type: "text" | "password";
  required?: boolean;
  valid?: boolean;
  errorInfo?: string;
  textInfo?: string;
  label?: string;
  disabled?: boolean;
  placeholder?: string;
}

const TextInput: React.FC<TextInputProps> = ({
  onChange,
  value,
  id,
  type,
  required = false,
  valid = true,
  errorInfo = "",
  textInfo = "",
  label = "",
  disabled = false,
  placeholder = "",
  className = "",
  testId = "",
  ...passThroughProps
}) => {
  return (
    <div className={className}>
      {label && (
        <label className="text-input-label" htmlFor={id}>
          {label} {required ? "(required)" : ""}
        </label>
      )}

      <input
        className={cn("text-input", { "text-input-error": !valid, "text-input-disabled": disabled })}
        value={value}
        id={id}
        onChange={onChange}
        type={type}
        placeholder={placeholder}
        disabled={disabled}
        data-testid={testId}
        aria-required={required}
        {...passThroughProps}
      />

      {valid && textInfo !== "" && <p className="text-input-info">{textInfo}</p>}

      {!valid && errorInfo !== "" && <p className="text-input-error-info">{errorInfo}</p>}
    </div>
  );
};

export default TextInput;
