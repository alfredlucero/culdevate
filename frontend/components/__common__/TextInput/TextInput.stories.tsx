import React from "react";
import { storiesOf } from "@storybook/react";
import StoryFrame from "../StoryFrame";
import TextInput, { TextInputType } from "./index";

const requiredTextInputProps = {
  value: "someTextInput",
  id: "some-text-input-id",
  name: "textInputName",
  valid: true,
  touched: true,
  type: "text" as TextInputType,
  onChange: () => {},
};

storiesOf("common/TextInput", module)
  .add("Default", () => (
    <StoryFrame>
      <TextInput {...requiredTextInputProps} />
      <TextInput {...requiredTextInputProps} value="some.email@domain.com" type="email" />
      <TextInput {...requiredTextInputProps} type="password" />
    </StoryFrame>
  ))
  .add("Valid + Info Message", () => (
    <StoryFrame>
      <TextInput {...requiredTextInputProps} infoMessage="Some info message" />
    </StoryFrame>
  ))
  .add("Error + Error Message", () => (
    <StoryFrame>
      <TextInput {...requiredTextInputProps} valid={false} errorMessage="Some error message" />
    </StoryFrame>
  ))
  .add("Label + placeholder + disabled + required", () => (
    <StoryFrame>
      <TextInput
        {...requiredTextInputProps}
        label="Some label"
        placeholder="Some placeholder text"
        value=""
        disabled={true}
        required={true}
      />
    </StoryFrame>
  ));
