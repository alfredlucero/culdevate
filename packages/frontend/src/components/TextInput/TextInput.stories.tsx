import React from "react";
import { storiesOf } from "@storybook/react";
import TextInput from "./index";

storiesOf("Common/Text Input", module)
  .add("Default Text", () => (
    <div style={{ width: "400px", padding: "30px" }}>
      <TextInput id="text-input" value="Text Value" onChange={() => {}} type="text" />
    </div>
  ))
  .add("Default Password", () => (
    <div style={{ width: "400px", padding: "30px" }}>
      <TextInput id="password-input" value="Password Value" onChange={() => {}} type="password" />
    </div>
  ))
  .add("Required Label", () => (
    <div style={{ width: "400px", padding: "30px" }}>
      <TextInput
        id="text-input"
        value="Required Username"
        onChange={() => {}}
        type="text"
        label="Username"
        required={true}
      />
    </div>
  ))
  .add("Disabled Placeholder", () => (
    <div style={{ width: "400px", padding: "30px" }}>
      <TextInput
        id="text-input"
        value=""
        placeholder="Enter a username..."
        onChange={() => {}}
        type="text"
        label="Username"
        disabled={true}
      />
    </div>
  ))
  .add("Inline Text Info", () => (
    <div style={{ width: "400px", padding: "30px" }}>
      <TextInput
        id="text-input"
        value="testuser123"
        onChange={() => {}}
        type="text"
        label="Username"
        valid={true}
        textInfo="Please provide a username."
        disabled={true}
      />
    </div>
  ))
  .add("Inline Text Info", () => (
    <div style={{ width: "400px", padding: "30px" }}>
      <TextInput
        id="text-input"
        value="testuser123"
        onChange={() => {}}
        type="text"
        label="Username"
        valid={true}
        textInfo="Please provide a username."
      />
    </div>
  ))
  .add("Inline Error Info", () => (
    <div style={{ width: "400px", padding: "30px" }}>
      <TextInput
        id="text-input"
        value="testuser123"
        onChange={() => {}}
        type="text"
        label="Username"
        valid={false}
        errorInfo="Please provide a username."
      />
    </div>
  ));
