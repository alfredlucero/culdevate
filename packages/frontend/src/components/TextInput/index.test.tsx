import React from "react";
import { render, fireEvent } from "@testing-library/react";
import TextInput from "./index";

describe("<TextInput />", () => {
  const inputTestId = "inputTestId";

  test("should render without error", () => {
    const { container } = render(
      <TextInput
        id="text-input-id"
        value="value"
        type="text"
        onChange={() => {}}
        testId={inputTestId}
        className="extra-input-class"
      />,
    );

    expect(container).toMatchSnapshot();
  });

  test("should render label if present", () => {
    const label = "Input Label";
    const { getByLabelText } = render(
      <TextInput id="text-input-id" value="value" type="text" onChange={() => {}} testId={inputTestId} label={label} />,
    );

    expect(getByLabelText(label)).toBeVisible();
  });

  test("should render required label if present and required", () => {
    const label = "Input Label";
    const { getByLabelText, getByTestId } = render(
      <TextInput
        id="text-input-id"
        value="value"
        type="text"
        onChange={() => {}}
        testId={inputTestId}
        label={label}
        required={true}
      />,
    );

    expect(getByLabelText(`${label} (required)`)).toBeVisible();
    expect(getByTestId(inputTestId)).toBeRequired();
  });

  test("should render inline text info if valid and info present", () => {
    const textInfo = "Some Text Info";
    const { container } = render(
      <TextInput
        id="text-input-id"
        value="value"
        type="text"
        onChange={() => {}}
        testId={inputTestId}
        textInfo={textInfo}
        errorInfo="Some error info"
        valid={true}
      />,
    );

    expect(container).toHaveTextContent(textInfo);
  });

  test("should render inline error info if invalid and error present", () => {
    const errorInfo = "Error Info";
    const { container, getByTestId } = render(
      <TextInput
        id="text-input-id"
        value="value"
        type="text"
        onChange={() => {}}
        testId={inputTestId}
        textInfo="Text Info"
        errorInfo={errorInfo}
        valid={false}
      />,
    );

    expect(container).toHaveTextContent(errorInfo);
  });

  test("should render placeholder given no value", () => {
    const placeholder = "Some placeholder...";
    const { getByPlaceholderText } = render(
      <TextInput
        id="text-input-id"
        value=""
        type="text"
        onChange={() => {}}
        testId={inputTestId}
        placeholder={placeholder}
      />,
    );

    expect(getByPlaceholderText(placeholder)).toBeVisible();
  });

  test("should render disabled state if disabled", () => {
    const { getByTestId } = render(
      <TextInput id="text-input-id" value="" type="text" onChange={() => {}} testId={inputTestId} disabled={true} />,
    );

    expect(getByTestId(inputTestId)).toBeDisabled();
  });

  test("should call onChange callback when firing change event", () => {
    const onChangeMock = jest.fn();
    const { getByTestId } = render(
      <TextInput
        id="text-input-id"
        value="value"
        type="text"
        onChange={onChangeMock}
        testId={inputTestId}
        className="extra-input-class"
      />,
    );

    fireEvent.change(getByTestId(inputTestId), { target: { value: "changedValue" } });

    expect(onChangeMock).toHaveBeenCalled();
  });
});
