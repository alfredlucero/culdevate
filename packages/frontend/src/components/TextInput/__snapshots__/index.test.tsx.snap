// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextInput /> should render without error 1`] = `
<div>
  <div
    class="extra-input-class"
  >
    
    <input
      aria-required="false"
      class="text-input"
      data-testid="inputTestId"
      id="text-input-id"
      placeholder=""
      type="text"
      value="value"
    />
  </div>
</div>
`;
